File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        currentNode=head
        nextNode=None
        previousNode=None
        while currentNode is not None:
            #nextNode holds next node
            nextNode=currentNode.next
            
            #connecting current node to previous node
            currentNode.next=previousNode
            
            #updating previous node to the current node
            previousNode=currentNode
            
            #updating current node to the next node
            currentNode= nextNode
            
        return previousNode
        

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode current=head;
        ListNode next=null;
        ListNode previous=null;
        while(current!=null){
            //next holds next node
            next=current.next;
            
            //connecting current node to previous node
            current.next=previous;
            
            //updating previous node to the current node
            previous=current;
            
            // updating current node to the next node
            current=next;
        }
        return previous;
    }
}

// [null,   1,     2,     3,     4,    5]
// next
//       current
// prev

// next=current.next
// current.next=prev